File: core/src/main/java/com/freactive/flawnkid/core/util/ShortcutReceiver.java
package com.freactive.flawnkid.core.util;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Parcelable;
import android.util.Log;

import com.freactive.flawnkid.core.R;
import com.freactive.flawnkid.core.activity.CoreHome;
import com.freactive.flawnkid.core.interfaces.AbstractApp;
import com.freactive.flawnkid.core.manager.Setup;
import com.freactive.flawnkid.core.model.Item;

public class ShortcutReceiver extends BroadcastReceiver {

    public ShortcutReceiver() {
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getExtras() == null) return;

        String name = intent.getExtras().getString(Intent.EXTRA_SHORTCUT_NAME);
        Intent newIntent = (Intent) intent.getExtras().get(Intent.EXTRA_SHORTCUT_INTENT);
        Drawable shortcutIconDrawable = null;

        try {
            Parcelable iconResourceParcelable = intent.getExtras().getParcelable(Intent.EXTRA_SHORTCUT_ICON_RESOURCE);
            if (iconResourceParcelable != null && iconResourceParcelable instanceof Intent.ShortcutIconResource) {
                Intent.ShortcutIconResource iconResource = (Intent.ShortcutIconResource) iconResourceParcelable;
                Resources resources = context.getPackageManager().getResourcesForApplication(iconResource.packageName);
                if (resources != null) {
                    int id = resources.getIdentifier(iconResource.resourceName, null, null);
                    shortcutIconDrawable = resources.getDrawable(id);
                }
            }
        } catch (Exception ignore) {
        } finally {
            if (shortcutIconDrawable == null)
                shortcutIconDrawable = new BitmapDrawable(context.getResources(), (Bitmap) intent.getExtras().getParcelable(Intent.EXTRA_SHORTCUT_ICON));
        }

        AbstractApp app = Setup.Companion.appLoader().createApp(newIntent);
        Item item;
        if (app != null) {
            item = Item.Companion.newAppItem(app);
        } else {
            item = Item.Companion.newShortcutItem(newIntent, shortcutIconDrawable, name);
        }
        Point preferredPos = CoreHome.Companion.getLauncher().getDesktop().getPages().get(CoreHome.Companion.getLauncher().getDesktop().getCurrentItem()).findFreeSpace();
        if (preferredPos == null) {
            Tool.INSTANCE.toast(CoreHome.Companion.getLauncher(), R.string.toast_not_enough_space);
        } else {
            item.setX(preferredPos.x);
            item.setY(preferredPos.y);
            CoreHome.Companion.getDb().saveItem(item, CoreHome.Companion.getLauncher().getDesktop().getCurrentItem(), Definitions.ItemPosition.Desktop);
            boolean added = CoreHome.Companion.getLauncher().getDesktop().addItemToPage(item, CoreHome.Companion.getLauncher().getDesktop().getCurrentItem());

            Setup.Companion.logger().log(this, Log.INFO, null, "Shortcut installed - %s => Intent: %s (Item type: %s; x = %d, y = %d, added = %b)", name, newIntent, item.getType(), item.getX(), item.getY(), added);
        }
    }
}

File: core/src/main/java/com/freactive/flawnkid/core/util/ShortcutReceiver.kt
package com.freactive.flawnkid.core.util

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Parcelable
import android.util.Log

import com.freactive.flawnkid.core.R
import com.freactive.flawnkid.core.activity.CoreHome
import com.freactive.flawnkid.core.interfaces.AbstractApp
import com.freactive.flawnkid.core.manager.Setup
import com.freactive.flawnkid.core.model.Item
import com.freactive.flawnkid.core.widget.CellContainer
import com.freactive.flawnkid.core.widget.OverviewPage

class ShortcutReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.extras == null) return

        val name = intent.extras!!.getString(Intent.EXTRA_SHORTCUT_NAME)
        val newIntent = intent.extras!!.get(Intent.EXTRA_SHORTCUT_INTENT) as Intent
        var shortcutIconDrawable: Drawable? = null

        try {
            val iconResourceParcelable = intent.extras!!.getParcelable<Parcelable>(Intent.EXTRA_SHORTCUT_ICON_RESOURCE)
            if (iconResourceParcelable != null && iconResourceParcelable is Intent.ShortcutIconResource) {
                val resources = context.packageManager.getResourcesForApplication(iconResourceParcelable.packageName)
                if (resources != null) {
                    val id = resources.getIdentifier(iconResourceParcelable.resourceName, null, null)
                    shortcutIconDrawable = resources.getDrawable(id)
                }
            }
        } catch (ignore: Exception) {
        } finally {
            if (shortcutIconDrawable == null)
                shortcutIconDrawable = BitmapDrawable(context.resources, intent.extras!!.getParcelable<Parcelable>(Intent.EXTRA_SHORTCUT_ICON) as Bitmap)
        }

        val app = Setup.appLoader().createApp(newIntent)
        val item: Item
        if (app != null) {
            item = Item.newAppItem(app)
        } else {
            item = Item.newShortcutItem(newIntent, shortcutIconDrawable, name!!)
        }
        val page = CoreHome.launcher!!.getDesktop().pages[CoreHome.launcher!!.getDesktop().currentItem]
        if (page is OverviewPage) return
            (page as CellContainer).apply {
            val preferredPos  = this.findFreeSpace()
            if (preferredPos == null) {
                Tool.toast(CoreHome.launcher!!, R.string.toast_not_enough_space)
            } else {
                item.x = preferredPos.x
                item.y = preferredPos.y
                CoreHome.db.saveItem(item, CoreHome.launcher!!.getDesktop().currentItem, Definitions.ItemPosition.Desktop)
                val added = CoreHome.launcher!!.getDesktop().addItemToPage(item, CoreHome.launcher!!.getDesktop().currentItem)

                //Setup.logger().log(this, Log.INFO, null, "Shortcut installed - %s => Intent: %s (Item type: %s; x = %d, y = %d, added = %b)", name, newIntent, item.type, item.x, item.y, added)
            }
        }
    }
}
